### ISSUE TITLE
[TI-96] refresh holidays in germany

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/timewarrior#100]
_Christian Decker on 2018-01-02T07:00:39Z says:_ 

 Hallo,

I've the following problem. When I make a refresh for holidays like in this link [https://taskwarrior.org/docs/timewarrior/holidays.html|https://taskwarrior.org/docs/timewarrior/holidays.html,] then I get an error because of the "ö" in Heilige drei Könige.

I think it's some kind of encoding error but I just got an workaround and no bugfix.
 
I switched 
json.loads(line)
to
json.loads(unicode(line, errors='ignore'))
 
Thanks for your help

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2018-01-02T07:00:39Z
Modified: 2018-01-15T05:43:06Z
```

**Comment by taskwarrior** [comment]
_Christian Decker on 2018-01-03T06:00:38Z says:_ 

 I solved the problem. Is there any possibility to make a pull request?

I just create the file in UTF-8 with the io.open instead of the open method and add the 'u' before every write.
{code}
1. !/usr/bin/env python
1. ###############################################################################
1. #
1. # Copyright 2006 - 2016, Paul Beckingham, Federico Hernandez.
1. #
1. # Permission is hereby granted, free of charge, to any person obtaining a copy
1. # of this software and associated documentation files (the "Software"), to deal
1. # in the Software without restriction, including without limitation the rights
1. # to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
1. # copies of the Software, and to permit persons to whom the Software is
1. # furnished to do so, subject to the following conditions:
1. #
1. # The above copyright notice and this permission notice shall be included
1. # in all copies or substantial portions of the Software.
1. #
1. # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
1. # OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
1. # FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
1. # THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
1. # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
1. # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
1. # SOFTWARE.
1. #
1. # http://www.opensource.org/licenses/mit-license.php
1. #
1. ###############################################################################

import os, sys, io, re, time, datetime, json, argparse, urllib2


def enumerate(path):
  if not os.path.exists(path):
    raise Exception("Directory '%s' does not exist" % path)

  found = []
  for path, dirs, files in os.walk(path, topdown=True, onerror=None, followlinks=False):
    found.extend(map(lambda x: os.path.join(path, x), files))
  return found


def holidata(locale, year):
  return "http://holidata.net/%s/%d.json" % (locale, year)


def update_locales(locales, regions):
  update_single_year(locales, regions, datetime.datetime.now().year)
  update_single_year(locales, regions, datetime.datetime.now().year + 1)


def update_single_year(locales, regions, year):
  for locale in locales:
    print holidata(locale, year)
    try:
      url = holidata(locale, year)
      lines = urllib2.urlopen(url).read()

      with io.open("holidays.%s" % locale, "w", encoding="utf-8") as fh:
        fh.write(u'# Holiday data provided by Holidata.net\n')
        fh.write(u'#   ' + url + '\n')
        fh.write(u'#   Generated ' + time.strftime('%Y-%m-%dT%H:%M:%S') + '\n\n')
        fh.write(u'define holidays:\n')
        fh.write(u'  ' + locale + ':\n')

        for line in lines.split('\n'):
          if line:
            j = json.loads(line)
            if j['region'] == '' or regions is None or len(regions) == 0 or j['region'] in regions:
              day = j['date'].replace("-", "_")
              desc = j['description']
              fh.write(u'    %s = %s\n' % (day, desc))


    except urllib2.HTTPError as e:
      if e.code == 404:
        print "Holidata.net does not have data for %s, for %d." % (locale, year)
      else:
        print e.code, e.read()


def main(args):
  if args.locale:
    update_locales(args.locale, args.region)
  else:
1.  Enumerate all holiday files in the current directory.
    locales = []
    re_holiday_file = re.compile("\/holidays.([a-z]{2}-[A-Z]{2}$)")
    for file in enumerate('.'):
      result = re_holiday_file.search(file)
      if result:
1.  Extract the locale name.
        locales.append(result.group(1))

    update_locales(locales, args.region)


if __name__ == "__main__":
  usage = """See http://holidata.net for details of supported locales and regions."""
  parser = argparse.ArgumentParser(description="Update holiday data files. Simply run 'refresh' to update all of them.")
  parser.add_argument('--locale', nargs='+', help='Specific locale to update.')
  parser.add_argument('--region', nargs='+', help='Specific locale region to update.')
  args = parser.parse_args()

  try:
    main(args)
  except Exception as msg:
    print 'Error:', msg

{code}

**Comment by taskwarrior** [comment]
_Christian Decker on 2018-01-15T05:43:06Z says:_ 

 Issue can be closed with version 1.1.0

**Comment by lauft** [comment]
Version 1.1.1 is out, closing issue
